volttron/platform/vip/healthservice.py

```diff
@@ -0,0 +1,131 @@
+# Battelle Memorial Institute. The views and opinions of authors expressed
+# herein do not necessarily state or reflect those of the
+# United States Government or any agency thereof.
+#
+# PACIFIC NORTHWEST NATIONAL LABORATORY operated by
+# BATTELLE for the UNITED STATES DEPARTMENT OF ENERGY
+# under Contract DE-AC05-76RL01830
+# }}}
+
+from collections import defaultdict
+from datetime import datetime
+import logging
+
+from volttron.platform.agent.known_identities import CONTROL_CONNECTION, PROCESS_IDENTITIES
+from volttron.platform.agent.utils import format_timestamp
+from volttron.platform.vip.agent import Agent, Core, RPC
+
+_log = logging.getLogger(__name__)
+
+
+class HealthService(Agent):
+
+    def __init__(self, **kwargs):
+        super(HealthService, self).__init__(**kwargs)
+
+        # Store the health stats for given peers in a dictionary with
+        # keys being the identity of the connected agent.
+        self._health_dict = defaultdict(dict)
+
+    def peer_added(self, peer):
+        """
+        The `peer_added` method should be called whenever an agent is connected to the
+        platform.
+
+        :param peer: The identity of the agent connected to the platform
+        """
+        self._health_dict[peer]['peer'] = peer
+        if not self._health_dict.get('service_agent'):
+            self._health_dict[peer]['service_agent'] = peer in PROCESS_IDENTITIES
+        self._health_dict[peer]['connected'] = format_timestamp(datetime.now())
+        self._health_dict[peer].pop('disconnected', None)
+
+    def peer_dropped(self, peer):
+        # TODO: Should there be an option for  a db/log file for agents coming and going from the platform?
+        self._health_dict[peer]['disconnected'] = format_timestamp(datetime.now())
+        del self._health_dict[peer]
+
+    @RPC.export
+    def get_platform_health(self):
+        """
+        The `get_platform_health` retrieves all of the connected agent's health structures,
+        except for the service_agents and the CONTROL_CONNECTION.  These are filtered out
+        because they aren't relevant to the vctl command in general.  The vctl command
+        usese CONTROL_CONNECTION to connect to the volttron process so it is not relevant
+        either.
+
+        This function returns a dictionary in the form identity: values such as the following:
+
+        .. code-block :: json
+
+            {
+                "listeneragent-3.3_35":
+                {
+                    "peer": "listeneragent-3.3_35",
+                    "service_agent": False,
+                    "connected": "2020-10-28T12:46:58.701119",
+                    "last_heartbeat": "2020-10-28T12:47:03.709605",
+                    "message": "GOOD"
+                }
+            }
+
+        :return:
+        """
+        # Ignore the connection from control in the health as it will only be around for a short while.
+        non_service_agents = {k: v for k, v in self._health_dict.items()
```

Comment:
'non_service_agents' variable name is incorrect since you are not filtering out service agents here